### DISCUSSION TITLE
Best Practices on testing different optimizers

### DISCUSSION THREAD

**Discussion by iglimlloja** [KellerJordan/modded-nanogpt#232] [Q&A]
I was trying to get my hands dirty with trying different optimizers while keeping the communication strategy that we have already (whenever possible), and thought that implementing NorAdam might be a good starting spot before getting into more unknown territories. Something is going terribly wrong somewhere along the way though, so I think it would be helpful to get some general direction/advice on how to do this best, and hopefully this might help others who might be running into similar issues.

My current understanding of the optimizer communication policy is as follows:

- First Pass: Each rank sends some gradient shard to be aggregated, i.e. each (replicated) model sends some part of the gradient estimated from their local micro batch, these are averaged and scattered.
- Second Pass: We wait for the grad shard aggregation to finish (and lazily initialize helper buffers in the case of NorMuon). We also compute the update in this pass (polar express, normuon variance reduction, cautious wd, parameter updates). Finally launch gather operation
- Third Pass: Copy results from the all gather into local model. They each have a shared view of the shards from all ranks, so model parameters across ranks are consistent.

To implement NorAdam, all I did was change the second pass, to lazily initialize the per-group EMA buffers (I don't think it is necessary for these to be per-parameter, as the group should remain fixed, but please correct me if I am mistaken), and to compute the Adam step to update the buffers and the update for that shard to be used in the all gather operation, while keeping third pass untouched.

Is this understanding correct?

If so, and in case this didn't get too tiring to read, any guesses on what could potentially lead to NaN gradients, even for zero step sizes?

**Comment by chrisjmccormick** [comment]
That all sounds correct except for maybe the "per-group EMA buffers" (normuon second momentum buffers).
These store one value per neuron, per matrix. 
Comparing it to Adam might clarify a bit--Adam stores a unique momentum value for every unique float in the weights. NorMuon is almost the same, except that it's one momentum value per neuron. (i.e., per row, or per column, of a weight matrix, depending on how it's oriented).
Does that help? I could be misunderstanding the question. :)

**Reply by iglimlloja** [reply]
I think this is a valuable clarification as what I said is kinda misleading, or imprecise at best. I think if I refer to the code directly it might be simplest to explain what I meant to say, and I guess it relates to the original question as well.

As you pointed out, in the case of NorMuon we are storing one value per neuron when we do `group["second_momentum_buffer"] = torch.zeros_like(updated_grads[..., :, :1])` (or per column instead `updated_grads[..., :1, :]`) by specifically passing the `:1` part. In implementing NorAdam, I simply added `group["exp_avg"] = torch.zeros_like(grad_chunk)`, and the same for `group["exp_avg_sq"]`, and changed polar express to the adam update, while keeping the cautious wd and second_momentum_buffer untouched. 

So I guess my original question (or perhaps thinking out loud effort) was about whether the assumption to make this _a group property_ -- in the case of 'NorAdam' -- instead of the traditional _param_state property_ is correct (to me it sounded like it should be unless I had misunderstood the communication flow). 

I guessed that my issue of NaN gradients was due to my implementation of NorAdam (in particular the assumption in the last sentence), which is why I wanted to double check whether my understanding of the communication policy was correct. Given that I seem to have correctly understood the communication part, either the assumption to make this a group property is wrong, in which case it would be great if someone could point out why as I've been scratching my head on this for longer than I'd like to admit lol, or I have introduced some silly bug elsewhere, which future me will love tracking.

**Comment by varunneal** [comment]
NaN gradients could honestly be caused by a lot of things, but it basically has to do with numerical imprecision or not being safe with certain operations like division. In my opinion, the most touchy part of the current training script has to do with the FP8 LM Head. If you drop the FP8 on the LM Head, you may find that the gradient no longer blows up.

**Reply by iglimlloja** [reply]
Forgot to update: I ended up cloning the commit with the merged NorMuon and Adam optimizer and was able to do what I was looking for without much trouble. I hope anyone who might be running into similar issues gives it a try with that setup in case they feel stuck.

I’m looking forward to getting some improvements so I can also join the leaderboard!
